### PR TITLE
server,cli,sql: enable setting a password for the root user

### LINKED ISSUE #43847: security: allow users to enable password auth for 'root'
tldr; this issue proposes to enable:

- setting/changing the password of the `root` user;
- allowing root to log in using their password on the UI;
- allowing root to log in using their password on SQL, subject to regular rules in the HBA configuration.

It also fixes the [regression](https://github.com/cockroachdb/cockroach/issues/43847#issuecomment-572615646)  introduced by #42563, by enabling `root` authentication on the admin UI to obtain a login cookie.

### Background

CockroachDB currently offers 5 separate guardrails to ensure that `root` is always able to connect even when the authentication configuration is botched:

1. when running `--insecure`, anyone can log in without auth.
2. the crdb [HBA configuration](https://www.postgresql.org/docs/current/auth-pg-hba-conf.html) enforces that `host all root all cert` is always the first rule, meaning that clients can log in as `root` if and only if they present a valid TLS client cert for `root` (and because the method is `cert` and not `cert-password`, password auth is rejected for `root` in any case).
3. `ALTER USER WITH PASSWORD` is disallowed for `root`, so root cannot have a password
4. the internal "check password" mechanism, shared by both SQL and HTTP connections, to compare a client-provided password with a user record fails with an error if the user is `root`.
5. the UI `UserLogin` HTTP API reports an error if the user is `root`.
6. the cockroach CLI commands using SQL connections report an error if the user is `root` and a password is supplied.

Meanwhile, there is a separate, unrelated (but relevant) rule:

7. if a user's password is unset/empty, then this user is not able to use password authentication.

### Proposal

This issue proposes to **remove rules 3-6**  specifically, without changing the others.

This proposal would not change the security rules for a cluster using the default configuration: by default, root would not be able to log in using password anywhere (by default, the root account has no password so rule 6 applies) and is required to present a cert on SQL (due to rule 2).

### Non-Pitfalls

A possible counter-argument to this proposal is that CockroachdB uses `root` internally to establish SQL client connections towards itself. 

This is not an obstacle: in insecure clusters, the password would be ignored anyway; in secure clusters, CockroachDB's internal connections use a valid client cert.

### PR DESCRIPTION
**Pull request by knz**
tldr: this patch enables the `root` user to have a password like
any user.

Fixes #43847. 

**Background**

Previously, CockroachDB implemented a plethora of special and highly
redundant cases to ensures that the `root` user could never have a
password and could never try to log in with a password, no matter
what. Only `root` was given these special cases, not other admin
users.

The somewhat reasonable first principle underlying the irrational
amount of redundant code is that contrary to most other DBs, it is
extremely difficult to re-add a valid user account into a cluster that
shut itself off by mis-configuring user accounts (short of re-starting
the entire cluster in insecure mode, which is arguably unpalatable in
production systems). It is desirable to guarantee `root` access
somehow. Allowing users to set, then forget, a `root` password
would endanger that.

This principle is not violated by this patch. This is described
further below.

Two additional principles were _originally_ valid but are not valid any
more:

- allowing users to configure passwords and enable password
  authentication provide a false sense of security, *in the particular
  case when the password is provided by a human having to memorize it
  and typing it on a keyboard.* [1] CockroachDB tries to
  promote industry best practices and promoting password
  authentication *with weak passwords* is not a best practice.

- CockroachDB "admin" users, like `root`, have the permission to
  pretty much do anything with a cluster, including viewing and
  modifying any data they want and hiding their traces. Therefore, admin
  users should be subject to a higher standard of scrutiny, and allowing
  them to have weak passwords (or any password at all, since
  CockroachDB has not entered the business of enforcing password
  hygiene) fails to match that higher standard.

This last principle is not valid any more because the `admin` role
can be granted to any user with any password, and it is accepted
to be industry-standard that password hygiene is the responsibility of
a site's operator, not that of the database engine.

[1] the password-on-keyboard case is the thing that's unsafe about
passwords. A sufficiently complex password entered non-interactively
by as password manager is virtually indistinguishable from a TLS client
certificate with regards to authentication.

Finally, CockroachDB does not support unencrypted network connections
in secure mode, so password sniffing on the network is not (currently)
possible. Allowing users to have and use passwords is not inherently
less secure than using and distributing TLS client certs, as transport
security is enforced using TLS in any case.

**Cluster safety consideration**

The previous section established that it is important that an admin
user _always_ be able to log in (currently, `root`) and that users are
not able to "shut themselves out" of operating their cluster.

This is currently implemented as follows in CockroachDB:

- in secure clusters, a valid TLS client for the `root` user is
  always guaranteed to yield a successful SQL authentication.
  This is in turn enforced by ensuring that the HBA configuration
  always start with a rule that allows cert auth for the `root` user.
- should the `root` cert be lost (and the CA key as well, preventing
  issuance of new `root` certs), it is possible to restart all the
  cluster with new CA whose key is known and issue a new `root` cert
  with that.

Remarkably, no mention is made here of passwords. This safety
guarantee is provided exclusively using TLS certs and the
special handling of HBA configurations.

**Description of changes**

This commit changes the following:

- removes the `root` special case in the UI login routine,
- removes the `root` special case in the
  `security.UserAuthPasswordHook` used to validate passwords
  through the various client channels,
- removes the `root` special case in the CLI client code that
  establishes SQL connections,
- removes the `root` special cases for password checks in CREATE USER
  and ALTER USER,
- changes the special HBA first rule to become `host all all root
  cert-password` instead of `host all all root cert`.

The commit in particular does not _remove_ the special HBA first rule
nor remove the ability to use a TLS client cert for `root`.

Additionally, the special case in `sql.GetUserHashedPassword` is
changed/evolved to enable root pw authentication, but still
guarantee that root can log in when some system range(s) are
unavailable. This is achieved as follows:

1) the interface of the function is changed to conceptually separate
   the lookup of whether the user exists from the lookup of their hashed
   password. This makes it possible to skip looking up the password
   when it is not needed.

2) by reporting that the root user exists without accessing
   system.users. This way, authentication methods that do not need a
   password (e.g. cert) do not need system.users to let root log in.

3) in case the root user is accessing via a password (e.g.  via the
   unix socket), by limiting the timeout to 5 seconds, so that large
   values of `system.user_login.timeout` do not impair the ability of
   operators to access the cluster.

Release note (security update): The `root` user is not anymore special
with regards to password management and can have a password like any
other member of the `admin` role. However, as in previous
versions, the HBA configuration cannot be overridden to prevent
`root` from logging in with a valid TLS client certificate. This
special rule remains enforced as a way to ensure that users cannot
"lock themselves out" of administrating their cluster.

Release note (security update): The `root` user remains special with
regards to authentication when some system ranges are unavailable: in
that case, password authentication will fail, subject to a timeout set
to the minimum between 5 seconds and the configured value of
`system.user_login_timeout` because the password cannot be retrieved
from a system table; however, certificate authentication remains
available.

### CODE REVIEW

File: pkg/cli/sql_util.go [hunk 1]
```diff
@@ -548,15 +548,7 @@ func makeSQLClient(appName string, defaultMode defaultSQLDb) (*sqlConn, error) {
...
-		if baseURL.User.Username() == security.RootUser {
-			// Disallow password login for root.
-			if options.Get("sslcert") == "" || options.Get("sslkey") == "" {
-				return nil, errors.Errorf("connections with user %s must use a client certificate",
-					baseURL.User.Username())
-			}
-			// If we can go on (we have a certificate spec), clear the password.
-			baseURL.User = url.User(security.RootUser)
-		} else if options.Get("sslcert") == "" || options.Get("sslkey") == "" {
+		if options.Get("sslcert") == "" || options.Get("sslkey") == "" {
```
Comment (mjibson) [inline pkg/cli/sql_util.go]: this if can be combined into the above else now

File: pkg/sql/pgwire/testdata/auth/password_change [hunk 2]
```diff
@@ -46,3 +48,34 @@ ok
...
+ALTER USER root WITH PASSWORD 'secureabc'
+----
+ok
+
+# Then they can log ini.
+connect user=root password=secureabc sslmode=require sslcert= sslkey=
+----
+ok defaultdb
+
+connect_unix user=root password=secureabc
```
Comment (mjibson) [inline pkg/sql/pgwire/testdata/auth/password_change]: Should we also test here connecting as root with a valid cert but incorrect password? That should work too right?

**Review by madelynnblue — APPROVED** [review]
LGTM, but since this is like auth policy stuff we should wait for both reviewers. I assume you may already have had that planned, but I'm just being explicit in the off chance it wasn't the case.

**Review by aaron-crl — COMMENTED** [review]
I like removing these scattered special cases around 'root' to standardize its auth patterns. 👍

(1) How do we want to communicate the changes to default hba conf behavior?
(2) Do the changes impact existing deployments or only new nodes?

**Comment by mjibson** [comment]
Also mixed version clusters during upgrades.

**Comment by knz** [comment]
> (1) How do we want to communicate the changes to default hba conf behavior?

I put a rather explanatory release note in there. This gets propagated in newsletter announcements too. I am hoping/expecting the docs team will expand this knowledge into a auth section in the ref docs.

> (2) Do the changes impact existing deployments or only new nodes?

Any node upgraded into 20.1 will adopt the new semantics. As Matt says that includes mixed 19.2/20.1 clusters.

**Comment by mjibson** [comment]
I just wanted to think through the mixed version scenario and figure out if security is decreased in any case.

1. On a node with the new version, the root user password is set. A user then tries to login as root with it on an old node. It says nope, root doesn't support passwords and fails the login. Not a security problem.
2. On a node with the new version, a user's password is set to NULL. That user tries to login on an old node. How does the old code handle the NULL there? Are the semantics safe on an old node for either the NULL or empty string case?
3. The HBA default (if it's empty) for root is `cert-password`. This just allows for more kinds of login, not less, so it's ok that old nodes continue to use `cert`.
4. If HBA is set to something on a new node, it gets the root `cert-password` auto-injected into its first line. If a user then tried to login to an old node as root (assuming the password had been set on a new node), the old node will attempt to use the password during auth, but the security/UserAuthPasswordHook function will disallow the attempt. This is fine.
5. I don't think old nodes can change something that wouldn't be forward compatible with a new node, so that's not a problem.

I think 2 is the only interesting question, but otherwise I don't think security is reduced in a mixed version cluster in any way.

**Review by aaron-crl — APPROVED** [review]
Sounds good to me.

**Comment by knz** [comment]
@mjibson 

> On a node with the new version, a user's password is set to NULL. That user tries to login on an old node. How does the old code handle the NULL there? Are the semantics safe on an old node for either the NULL or empty string case?

1) "password is set to NULL" = "password is removed"

That's also what happens initially when you do `CREATE USER foo` without a password, then the password does not exist.

In fact, it was previously possible to clear a password by doing `DROP USER foo; CREATE USER foo`.

So here you have your answer: the old node understands "no password" as "cannot log in". We have ample tests for this.

2) we do not allow using empty strings as passwords. That's enforced since forever in CREATE USER / ALTER USER.

**Comment by knz** [comment]
(FYI the remaining roachtest failure is due to related issue #43974. I'll look at fixing that before merging this)

**Comment by knz** [comment]
@mjibson I added a cluster setting to gate setting the password, to clarify the behavior in mixed versions.

**Review by madelynnblue — APPROVED** [review]

**Comment by knz** [comment]
bors r=mjibson,aaron-crl

**Comment by knz** [comment]
bors r=mjibson,aaron-crl

**Comment by knz** [comment]
this week's merge gods are against us

bors r=mjibson,aaron-crl
